### PR TITLE
Add initial expression editor workspace support

### LINKED ISSUE #33336: Add LS compilation support for cloned projects with dedicated file scheme
**Description:**
$title
Consider the use-case where an expression editor implementation wants to consume the language server APIs to consume features such as auto-completion and etc. 
In these cases, the LS should be able to re-use the existing project as a clone and provide context aware language support. In order to do so,

We can introduce a separate file scheme, say `expr` and then save a copy on the LS project cache under the given project path with the new scheme.

### PR DESCRIPTION
**Pull request by nadeeshaan**
## Purpose
Fixes #33336 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaWorkspaceManagerProxy.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+/**
+ * Ballerina workspace manager proxy implementation.
+ * This proxy maintains two workspace managers, one for the expr file scheme based documents and the default manager
+ * for the file scheme based documents.
+ *
+ * @since 2.0.0
+ */
+public class BallerinaWorkspaceManagerProxy implements WorkspaceManagerProxy {
+    private static final String EXPR_SCHEME = "expr";
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaWorkspaceManagerProxy.java]: We have this constant defined in CommonUtil as well.

**Review by IMS94 — CHANGES_REQUESTED** [review]
Changes look good to me. There's a minor comment. Will go through this once again later.
Btw, what's the plan to test this? Are we going to ask vscode plugin team to build your branch (jballerina tools) and test it with their version of the plugin? Or directly merge this?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaBallerinaWorkspaceManagerProxyImpl.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+import java.util.Optional;
+
+/**
+ * Ballerina workspace manager proxy implementation.
+ * This proxy maintains two workspace managers, one for the expr file scheme based documents and the default manager
+ * for the file scheme based documents.
+ *
+ * @since 2.0.0
+ */
+public class BallerinaBallerinaWorkspaceManagerProxyImpl implements BallerinaWorkspaceManagerProxy {
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaBallerinaWorkspaceManagerProxyImpl.java:40]: Was this name given purposefully? (2 "ballerina" words?)

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/diagnostic/DiagnosticsHelper.java [hunk 3]
```diff
@@ -151,9 +152,9 @@ public synchronized void compileAndSendDiagnostics(ExtendedLanguageClient client
      * @param compilation package compilation
      */
     private synchronized void compileAndSendDiagnostics(ExtendedLanguageClient client, Path projectRoot,
-                                                        PackageCompilation compilation) {
+                                                        PackageCompilation compilation, WorkspaceManager workspace) {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/diagnostic/DiagnosticsHelper.java:155]: Better if we can change `workspace` to `workspaceManager`

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/LSClientUtil.java [hunk 4]
```diff
@@ -150,6 +150,11 @@ public static boolean isDynamicSynchronizationRegistrationSupported(TextDocument
                 Boolean.TRUE.equals(capabilities.getSynchronization().getDynamicRegistration());
     }
 
+    public static boolean isDynamicCompletionRegistrationSupported(TextDocumentClientCapabilities capabilities) {
+        return capabilities != null && capabilities.getSynchronization() != null &&
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/LSClientUtil.java:154]:
```suggestion
        return capabilities != null && capabilities.getCompletion() != null &&
```

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/TestUtil.java [hunk 5]
```diff
@@ -571,13 +601,21 @@ public static String getWorkspaceSymbolResponse(Endpoint serviceEndpoint, String
         return getResponseString(result);
     }
 
+    @Deprecated
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/TestUtil.java:604]: Shall we add the deprecated documentation asking the user to use the new method

**Review by malinthar — CHANGES_REQUESTED** [review]
Implementation looks good.I have put a minor comment and a change request.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/TestUtil.java (hunk 5 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/TestUtil.java:610]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/LSClientUtil.java (hunk 4 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/LSClientUtil.java]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/diagnostic/DiagnosticsHelper.java (hunk 3 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/diagnostic/DiagnosticsHelper.java]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaWorkspaceManagerProxy.java (hunk 1 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaWorkspaceManagerProxy.java]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaBallerinaWorkspaceManagerProxyImpl.java (hunk 2 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaBallerinaWorkspaceManagerProxyImpl.java]: Fixed

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/TestUtil.java [hunk 6]
```diff
@@ -460,22 +477,35 @@ public static String getSemanticTokensResponse(Endpoint serviceEndpoint, String
...
+    }
+
+    /**
+     * Open a document.
+     *
+     * @param serviceEndpoint Language Server Service Endpoint
+     * @param fileUri         uri of the document to open
+     * @throws IOException Exception while reading the file content
+     */
+    public static void openDocument(Endpoint serviceEndpoint, String fileUri, String content) throws IOException {
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/TestUtil.java:491]: Missing description for `@param content`

**Review by IMS94 — APPROVED** [review]
LGTM!

**Review by malinthar — APPROVED** [review]
LGTM!

**Pull request merged**